File: nano/core_test/network.cpp
#include <nano/lib/blocks.hpp>
#include <nano/lib/config.hpp>
#include <nano/node/network.hpp>
#include <nano/node/nodeconfig.hpp>
#include <nano/node/scheduler/component.hpp>
#include <nano/node/scheduler/priority.hpp>
#include <nano/node/transport/tcp_listener.hpp>
#include <nano/secure/ledger.hpp>
#include <nano/store/component.hpp>
#include <nano/test_common/network.hpp>
#include <nano/test_common/system.hpp>
#include <nano/test_common/testutil.hpp>

#include <gtest/gtest.h>

#include <boost/iostreams/stream_buffer.hpp>
#include <boost/range/join.hpp>
#include <boost/thread.hpp>

#include <thread>

using namespace std::chrono_literals;

TEST (network, multi_keepalive)
{
	nano::test::system system (1);
	auto node0 = system.nodes[0];
	ASSERT_EQ (0, node0->network->size ());
	auto node1 (std::make_shared<nano::node> (system.async_rt, system.get_available_port (), nano::unique_path (), system.work));
	ASSERT_FALSE (node1->init_error ());
	node1->start ();
	system.nodes.push_back (node1);
	ASSERT_EQ (0, node1->network->size ());
	ASSERT_EQ (0, node0->network->size ());
	node1->connect (node0->network->endpoint ());
	ASSERT_TIMELY (10s, node0->network->size () == 1 && node0->stats->count (nano::stat::type::message, nano::stat::detail::keepalive) >= 1);
	auto node2 (std::make_shared<nano::node> (system.async_rt, system.get_available_port (), nano::unique_path (), system.work));
	ASSERT_FALSE (node2->init_error ());
	node2->start ();
	system.nodes.push_back (node2);
	node2->connect (node0->network->endpoint ());
	// ASSERT_TIMELY (10s, node1->network->size () == 2 && node0->network->size () == 2 && node2->network->size () == 2 && node0->stats->count (nano::stat::type::message, nano::stat::detail::keepalive) >= 2);
	std::this_thread::sleep_for (10s);
	std::cout << "node0: " << node0->network->size () << ", node1: " << node1->network->size () << ", node2: " << node2->network->size () << std::endl;
}

TEST (network, send_valid_confirm_ack)
{
	nano::node_flags node_flags;
	nano::test::system system (2, node_flags);
	auto & node1 (*system.nodes[0]);
	auto & node2 (*system.nodes[1]);
	auto wallet_id1 = node1.wallets.first_wallet_id ();
	auto wallet_id2 = node2.wallets.first_wallet_id ();
	nano::keypair key2;
	(void)node1.wallets.insert_adhoc (wallet_id1, nano::dev::genesis_key.prv);
	(void)node2.wallets.insert_adhoc (wallet_id2, key2.prv);
	nano::block_hash latest1 (node1.latest (nano::dev::genesis_key.pub));
	nano::block_builder builder;
	auto block2 = builder
				  .send ()
				  .previous (latest1)
				  .destination (key2.pub)
				  .balance (50)
				  .sign (nano::dev::genesis_key.prv, nano::dev::genesis_key.pub)
				  .work (*system.work.generate (latest1))
				  .build ();
	nano::block_hash latest2 (node2.latest (nano::dev::genesis_key.pub));
	node1.process_active (std::make_shared<nano::send_block> (*block2));
	// Keep polling until latest block changes
	ASSERT_TIMELY (10s, node2.latest (nano::dev::genesis_key.pub) != latest2);
	// Make sure the balance has decreased after processing the block.
	ASSERT_EQ (50, node2.balance (nano::dev::genesis_key.pub));
}

TEST (network, send_valid_publish)
{
	nano::node_flags node_flags;
	nano::test::system system (2, node_flags);
	auto & node1 (*system.nodes[0]);
	auto & node2 (*system.nodes[1]);
	auto wallet_id1 = node1.wallets.first_wallet_id ();
	auto wallet_id2 = node2.wallets.first_wallet_id ();
	node1.bootstrap_initiator.stop ();
	node2.bootstrap_initiator.stop ();
	(void)node1.wallets.insert_adhoc (wallet_id1, nano::dev::genesis_key.prv);
	nano::keypair key2;
	(void)node2.wallets.insert_adhoc (wallet_id2, key2.prv);
	nano::block_hash latest1 (node1.latest (nano::dev::genesis_key.pub));
	nano::block_builder builder;
	auto block2 = builder
				  .send ()
				  .previous (latest1)
				  .destination (key2.pub)
				  .balance (50)
				  .sign (nano::dev::genesis_key.prv, nano::dev::genesis_key.pub)
				  .work (*system.work.generate (latest1))
				  .build ();
	auto hash2 (block2->hash ());
	nano::block_hash latest2 (node2.latest (nano::dev::genesis_key.pub));
	node2.process_active (std::make_shared<nano::send_block> (*block2));
	ASSERT_TIMELY (10s, node1.stats->count (nano::stat::type::message, nano::stat::detail::publish, nano::stat::dir::in) != 0);
	ASSERT_NE (hash2, latest2);
	ASSERT_TIMELY (10s, node2.latest (nano::dev::genesis_key.pub) != latest2);
	ASSERT_EQ (50, node2.balance (nano::dev::genesis_key.pub));
}

TEST (receivable_processor, send_with_receive)
{
	nano::node_flags node_flags;
	nano::test::system system (2, node_flags);
	auto & node1 (*system.nodes[0]);
	auto & node2 (*system.nodes[1]);
	auto wallet_id1 = node1.wallets.first_wallet_id ();
	auto wallet_id2 = node2.wallets.first_wallet_id ();
	auto amount (std::numeric_limits<nano::uint128_t>::max ());
	nano::keypair key2;
	(void)node1.wallets.insert_adhoc (wallet_id1, nano::dev::genesis_key.prv);
	nano::block_hash latest1 (node1.latest (nano::dev::genesis_key.pub));
	nano::block_builder builder;
	auto block1 = builder
				  .send ()
				  .previous (latest1)
				  .destination (key2.pub)
				  .balance (amount - node1.config->receive_minimum.number ())
				  .sign (nano::dev::genesis_key.prv, nano::dev::genesis_key.pub)
				  .work (*system.work.generate (latest1))
				  .build ();
	ASSERT_EQ (amount, node1.balance (nano::dev::genesis_key.pub));
	ASSERT_EQ (0, node1.balance (key2.pub));
	ASSERT_EQ (amount, node2.balance (nano::dev::genesis_key.pub));
	ASSERT_EQ (0, node2.balance (key2.pub));
	node1.process_active (block1);
	ASSERT_TIMELY (5s, nano::test::exists (node1, { block1 }));
	node2.process_active (block1);
	ASSERT_TIMELY (5s, nano::test::exists (node2, { block1 }));
	ASSERT_EQ (amount - node1.config->receive_minimum.number (), node1.balance (nano::dev::genesis_key.pub));
	ASSERT_EQ (0, node1.balance (key2.pub));
	ASSERT_EQ (amount - node1.config->receive_minimum.number (), node2.balance (nano::dev::genesis_key.pub));
	ASSERT_EQ (0, node2.balance (key2.pub));
	(void)node2.wallets.insert_adhoc (wallet_id2, key2.prv);
	ASSERT_TIMELY (10s, node1.balance (key2.pub) == node1.config->receive_minimum.number () && node2.balance (key2.pub) == node1.config->receive_minimum.number ());
	ASSERT_EQ (amount - node1.config->receive_minimum.number (), node1.balance (nano::dev::genesis_key.pub));
	ASSERT_EQ (node1.config->receive_minimum.number (), node1.balance (key2.pub));
	ASSERT_EQ (amount - node1.config->receive_minimum.number (), node2.balance (nano::dev::genesis_key.pub));
	ASSERT_EQ (node1.config->receive_minimum.number (), node2.balance (key2.pub));
}

TEST (network, receive_weight_change)
{
	nano::test::system system (2);
	auto node1 = system.nodes[0];
	auto node2 = system.nodes[1];
	auto wallet_id1 = node1->wallets.first_wallet_id ();
	auto wallet_id2 = node2->wallets.first_wallet_id ();
	(void)node1->wallets.insert_adhoc (wallet_id1, nano::dev::genesis_key.prv);
	nano::keypair key2;
	(void)node2->wallets.insert_adhoc (wallet_id2, key2.prv);
	(void)node2->wallets.set_representative (wallet_id2, key2.pub);
	ASSERT_NE (nullptr, node1->wallets.send_action (wallet_id1, nano::dev::genesis_key.pub, key2.pub, system.nodes[0]->config->receive_minimum.number ()));
	ASSERT_TIMELY (10s, std::all_of (system.nodes.begin (), system.nodes.end (), [&] (std::shared_ptr<nano::node> const & node_a) { return node_a->weight (key2.pub) == system.nodes[0]->config->receive_minimum.number (); }));
}

File: node/tests/tests/network.rs
use rsnano_core::{
    Account, Amount, Block, KeyPair, Networks, Root, StateBlock, Vote, DEV_GENESIS_KEY,
};
use rsnano_ledger::{DEV_GENESIS_ACCOUNT, DEV_GENESIS_HASH, DEV_GENESIS_PUB_KEY};
use rsnano_messages::{
    ConfirmAck, Keepalive, Message, MessageHeader, MessageSerializer, ProtocolInfo, Publish,
};
use rsnano_network::{ChannelMode, DropPolicy, TrafficType};
use rsnano_node::{
    bootstrap::BootstrapInitiatorExt,
    config::NodeConfig,
    consensus::VoteProcessorConfig,
    stats::{DetailType, Direction, StatType},
    wallets::WalletsExt,
    Node,
};
use std::{ops::Deref, sync::Arc, thread::sleep, time::Duration};
use test_helpers::{
    assert_always_eq, assert_timely, assert_timely_eq, assert_timely_msg, establish_tcp,
    make_fake_channel, start_election, System,
};

#[test]
fn last_contacted() {
    let mut system = System::new();

    let node0 = system.make_node();

    let mut node1_config = System::default_config();
    node1_config.tcp_incoming_connections_max = 0; // Prevent ephemeral node1->node0 channel repacement with incoming connection
    let node1 = system
        .build_node()
        .config(node1_config)
        .disconnected()
        .finish();

    let channel1 = establish_tcp(&node1, &node0);
    assert_timely_eq(
        Duration::from_secs(3),
        || {
            node0
                .network_info
                .read()
                .unwrap()
                .count_by_mode(ChannelMode::Realtime)
        },
        1,
    );

    // channel0 is the other side of channel1, same connection different endpoint
    let channel0 = node0
        .network_info
        .read()
        .unwrap()
        .find_node_id(&node1.node_id.public_key())
        .unwrap()
        .clone();

    // check that the endpoints are part of the same connection
    assert_eq!(channel0.local_addr(), channel1.peer_addr());
    assert_eq!(channel1.local_addr(), channel0.peer_addr());

    // capture the state before and ensure the clock ticks at least once
    let timestamp_before_keepalive = channel0.last_activity();
    let keepalive_count =
        node0
            .stats
            .count(StatType::Message, DetailType::Keepalive, Direction::In);
    assert_timely_msg(
        Duration::from_secs(3),
        || node0.steady_clock.now() > timestamp_before_keepalive,
        "clock did not advance",
    );

    // send 3 keepalives
    // we need an extra keepalive to handle the race condition between the timestamp set and the counter increment
    // and we need one more keepalive to handle the possibility that there is a keepalive already in flight when we start the crucial part of the test
    // it is possible that there could be multiple keepalives in flight but we assume here that there will be no more than one in flight for the purposes of this test
    let keepalive = Message::Keepalive(Keepalive::default());
    let mut publisher = node0.message_publisher.lock().unwrap();
    publisher.try_send(
        channel1.channel_id(),
        &keepalive,
        DropPolicy::ShouldNotDrop,
        TrafficType::Generic,
    );
    publisher.try_send(
        channel1.channel_id(),
        &keepalive,
        DropPolicy::ShouldNotDrop,
        TrafficType::Generic,
    );
    publisher.try_send(
        channel1.channel_id(),
        &keepalive,
        DropPolicy::ShouldNotDrop,
        TrafficType::Generic,
    );

    assert_timely_msg(
        Duration::from_secs(3),
        || {
            node0
                .stats
                .count(StatType::Message, DetailType::Keepalive, Direction::In)
                >= keepalive_count + 3
        },
        "keepalive count",
    );
    assert_eq!(
        node0
            .network_info
            .read()
            .unwrap()
            .count_by_mode(ChannelMode::Realtime),
        1
    );
    let timestamp_after_keepalive = channel0.last_activity();
    assert!(timestamp_after_keepalive > timestamp_before_keepalive);
}

#[test]
fn send_discarded_publish() {
    let mut system = System::new();
    let node1 = system.make_node();
    let node2 = system.make_node();

    let block = Block::State(StateBlock::new(
        *DEV_GENESIS_ACCOUNT,
        2.into(),
        3.into(),
        Amount::MAX,
        4.into(),
        &DEV_GENESIS_KEY,
        node1.work_generate_dev(Root::from(2)),
    ));

    node1.message_publisher.lock().unwrap().flood(
        &Message::Publish(Publish::new_forward(block)),
        DropPolicy::ShouldNotDrop,
        1.0,
    );

    assert_eq!(node1.latest(&DEV_GENESIS_ACCOUNT), *DEV_GENESIS_HASH);
    assert_eq!(node2.latest(&DEV_GENESIS_ACCOUNT), *DEV_GENESIS_HASH);
    assert_timely_msg(
        Duration::from_secs(10),
        || {
            node2
                .stats
                .count(StatType::Message, DetailType::Publish, Direction::In)
                != 0
        },
        "no publish received",
    );
    assert_eq!(node1.latest(&DEV_GENESIS_ACCOUNT), *DEV_GENESIS_HASH);
    assert_eq!(node2.latest(&DEV_GENESIS_ACCOUNT), *DEV_GENESIS_HASH);
}

#[test]
fn receivable_processor_confirm_insufficient_pos() {
    let mut system = System::new();
    let node1 = system.make_node();
    let send1 = Block::State(StateBlock::new(
        *DEV_GENESIS_ACCOUNT,
        *DEV_GENESIS_HASH,
        *DEV_GENESIS_PUB_KEY,
        Amount::MAX - Amount::raw(1),
        Account::zero().into(),
        &DEV_GENESIS_KEY,
        node1.work_generate_dev(*DEV_GENESIS_HASH),
    ));

    node1.process(send1.clone()).unwrap();
    let election = start_election(&node1, &send1.hash());

    let key1 = KeyPair::new();
    let vote = Arc::new(Vote::new_final(&key1, vec![send1.hash()]));
    let channel = make_fake_channel(&node1);
    let con1 = Message::ConfirmAck(ConfirmAck::new_with_rebroadcasted_vote(
        vote.deref().clone(),
    ));
    assert_eq!(1, election.vote_count());

    node1.inbound_message_queue.put(con1, channel.info.clone());

    assert_timely_eq(Duration::from_secs(5), || election.vote_count(), 2);
}

#[test]
fn receivable_processor_confirm_sufficient_pos() {
    let mut system = System::new();
    let node1 = system.make_node();
    let send1 = Block::State(StateBlock::new(
        *DEV_GENESIS_ACCOUNT,
        *DEV_GENESIS_HASH,
        *DEV_GENESIS_PUB_KEY,
        Amount::MAX - Amount::raw(1),
        Account::zero().into(),
        &DEV_GENESIS_KEY,
        node1.work_generate_dev(*DEV_GENESIS_HASH),
    ));

    node1.process(send1.clone()).unwrap();
    let election = start_election(&node1, &send1.hash());

    let vote = Arc::new(Vote::new_final(&DEV_GENESIS_KEY, vec![send1.hash()]));
    let channel = make_fake_channel(&node1);
    let con1 = Message::ConfirmAck(ConfirmAck::new_with_rebroadcasted_vote(
        vote.deref().clone(),
    ));
    assert_eq!(1, election.vote_count());

    node1.inbound_message_queue.put(con1, channel.info.clone());

    assert_timely_eq(Duration::from_secs(5), || election.vote_count(), 2);
}

#[test]
fn multi_keepalive() {
    let mut system = System::new();
    let node1 = system.make_node();
    let _node2 = system.make_node();
    assert_timely(Duration::from_secs(10), || {
        node1
            .stats
            .count(StatType::Message, DetailType::Keepalive, Direction::In)
            > 0
    });
}

#[test]
fn send_valid_confirm_ack() {
    let mut system = System::new();
    let node1 = system.make_node();
    let node2 = system.make_node();
    let key2 = KeyPair::new();
    node1.insert_into_wallet(&DEV_GENESIS_KEY);
    node2.insert_into_wallet(&key2);
    let block2 = Block::State(StateBlock::new(
        *DEV_GENESIS_ACCOUNT,
        *DEV_GENESIS_HASH,
        *DEV_GENESIS_PUB_KEY,
        Amount::raw(50),
        key2.public_key().as_account().into(),
        &DEV_GENESIS_KEY,
        node1.work_generate_dev(*DEV_GENESIS_HASH),
    ));
    node1.process_active(block2);
    // Keep polling until latest block changes
    assert_timely(Duration::from_secs(10), || {
        node2.latest(&DEV_GENESIS_ACCOUNT) != *DEV_GENESIS_HASH
    });
    // Make sure the balance has decreased after processing the block.
    assert_eq!(node2.balance(&DEV_GENESIS_ACCOUNT), Amount::raw(50));
}

#[test]
fn send_valid_publish() {
    let mut system = System::new();
    let node1 = system.make_node();
    let node2 = system.make_node();
    node1.bootstrap_initiator.stop();
    node2.bootstrap_initiator.stop();
    node1.insert_into_wallet(&DEV_GENESIS_KEY);
    let key2 = KeyPair::new();
    node2.insert_into_wallet(&key2);
    let block2 = Block::State(StateBlock::new(
        *DEV_GENESIS_ACCOUNT,
        *DEV_GENESIS_HASH,
        *DEV_GENESIS_PUB_KEY,
        Amount::raw(50),
        key2.public_key().as_account().into(),
        &DEV_GENESIS_KEY,
        node1.work_generate_dev(*DEV_GENESIS_HASH),
    ));
    let hash2 = block2.hash();
    let latest2 = node2.latest(&DEV_GENESIS_ACCOUNT);
    node2.process_active(block2);
    assert_timely(Duration::from_secs(10), || {
        node1
            .stats
            .count(StatType::Message, DetailType::Publish, Direction::In)
            > 0
    });
    assert_ne!(hash2, latest2);
    assert_timely(Duration::from_secs(10), || {
        node2.latest(&DEV_GENESIS_ACCOUNT) != latest2
    });
    assert_eq!(node2.balance(&DEV_GENESIS_ACCOUNT), Amount::raw(50));
}

#[test]
fn send_with_receive() {
    let mut system = System::new();
    let node1 = system.make_node();
    let node2 = system.make_node();
    let key2 = KeyPair::new();
    node1.insert_into_wallet(&DEV_GENESIS_KEY);
    let block1 = Block::State(StateBlock::new(
        *DEV_GENESIS_ACCOUNT,
        *DEV_GENESIS_HASH,
        *DEV_GENESIS_PUB_KEY,
        Amount::MAX - node1.config.receive_minimum,
        key2.public_key().as_account().into(),
        &DEV_GENESIS_KEY,
        node1.work_generate_dev(*DEV_GENESIS_HASH),
    ));

    node1.process_active(block1.clone());
    assert_timely(Duration::from_secs(5), || {
        node1.block_exists(&block1.hash())
    });
    node2.process_active(block1.clone());
    assert_timely(Duration::from_secs(5), || {
        node2.block_exists(&block1.hash())
    });
    node2.insert_into_wallet(&key2);
    assert_timely(Duration::from_secs(10), || {
        node1.balance(&key2.public_key().as_account()) == node1.config.receive_minimum
            && node2.balance(&key2.public_key().as_account()) == node1.config.receive_minimum
    });
}

#[test]
fn receive_weight_change() {
    let mut system = System::new();
    let node1 = system.make_node();
    let node2 = system.make_node();
    let key2 = KeyPair::new();
    node1.insert_into_wallet(&DEV_GENESIS_KEY);
    node2.insert_into_wallet(&key2);
    node2
        .wallets
        .set_representative(node2.wallets.wallet_ids()[0], key2.public_key(), false)
        .unwrap();
    node1
        .wallets
        .send_action2(
            &node1.wallets.wallet_ids()[0],
            *DEV_GENESIS_ACCOUNT,
            key2.public_key().as_account(),
            node1.config.receive_minimum,
            0,
            true,
            None,
        )
        .unwrap();
    assert_timely(Duration::from_secs(10), || {
        node1
            .ledger
            .weight_exact(&node1.ledger.read_txn(), key2.public_key())
            == node1.config.receive_minimum
            && node2
                .ledger
                .weight_exact(&node2.ledger.read_txn(), key2.public_key())
                == node1.config.receive_minimum
    });
}

#[test]
fn duplicate_vote_detection() {
    let mut system = System::new();
    let node0 = system.make_node();
    let node1 = system.make_node();

    let vote = Vote::new(&DEV_GENESIS_KEY, 0, 0, vec![*DEV_GENESIS_HASH]);
    let message = Message::ConfirmAck(ConfirmAck::new_with_own_vote(vote));

    // Publish duplicate detection through TCP
    let channel_id = node0
        .network_info
        .read()
        .unwrap()
        .find_node_id(&node1.node_id.public_key())
        .unwrap()
        .channel_id();

    node0.message_publisher.lock().unwrap().try_send(
        channel_id,
        &message,
        DropPolicy::ShouldNotDrop,
        TrafficType::Generic,
    );
    assert_always_eq(
        Duration::from_millis(100),
        || {
            node1.stats.count(
                StatType::Filter,
                DetailType::DuplicateConfirmAckMessage,
                Direction::In,
            )
        },
        0,
    );
    node0.message_publisher.lock().unwrap().try_send(
        channel_id,
        &message,
        DropPolicy::ShouldNotDrop,
        TrafficType::Generic,
    );
    assert_timely_eq(
        Duration::from_secs(2),
        || {
            node1.stats.count(
                StatType::Filter,
                DetailType::DuplicateConfirmAckMessage,
                Direction::In,
            )
        },
        1,
    );
}

// Ensures that the filter doesn't filter out votes that could not be queued for processing
#[test]
fn duplicate_revert_vote() {
    let mut system = System::new();
    let node0 = system
        .build_node()
        .config(NodeConfig {
            enable_vote_processor: false, // do not drain queued votes
            vote_processor: VoteProcessorConfig {
                max_non_pr_queue: 1,
                max_pr_queue: 1,
                ..VoteProcessorConfig::new(1)
            },
            ..System::default_config()
        })
        .finish();
    let node1 = system
        .build_node()
        .config(NodeConfig {
            enable_vote_processor: false,
            vote_processor: VoteProcessorConfig {
                max_non_pr_queue: 1,
                max_pr_queue: 1,
                ..VoteProcessorConfig::new(1)
            },
            ..System::default_config()
        })
        .finish();

    let vote1 = Vote::new(&DEV_GENESIS_KEY, 1, 0, vec![*DEV_GENESIS_HASH]);
    let message1 = Message::ConfirmAck(ConfirmAck::new_with_own_vote(vote1));

    let vote2 = Vote::new(&DEV_GENESIS_KEY, 2, 2, vec![*DEV_GENESIS_HASH]);
    let message2 = Message::ConfirmAck(ConfirmAck::new_with_own_vote(vote2));

    // Publish duplicate detection through TCP
    let channel_id = node0
        .network_info
        .read()
        .unwrap()
        .find_node_id(&node1.node_id.public_key())
        .unwrap()
        .channel_id();

    // First vote should be processed
    node0.message_publisher.lock().unwrap().try_send(
        channel_id,
        &message1,
        DropPolicy::ShouldNotDrop,
        TrafficType::Generic,
    );
    assert_always_eq(
        Duration::from_millis(100),
        || {
            node1.stats.count(
                StatType::Filter,
                DetailType::DuplicateConfirmAckMessage,
                Direction::In,
            )
        },
        0,
    );

    // Second vote should get dropped from processor queue
    node0.message_publisher.lock().unwrap().try_send(
        channel_id,
        &message2,
        DropPolicy::ShouldNotDrop,
        TrafficType::Generic,
    );
    assert_always_eq(
        Duration::from_millis(100),
        || {
            node1.stats.count(
                StatType::Filter,
                DetailType::DuplicateConfirmAckMessage,
                Direction::In,
            )
        },
        0,
    );
    // And the filter should not have it
    sleep(Duration::from_millis(500)); // Give the node time to process the vote

    let mut serializer =
        MessageSerializer::new(ProtocolInfo::default_for(Networks::NanoDevNetwork));
    let msg2_bytes = serializer.serialize(&message2);
    let payload_bytes = &msg2_bytes[MessageHeader::SERIALIZED_SIZE..];
    assert_eq!(node1.network_filter.check_message(payload_bytes), false);
}

#[test]
fn expire_duplicate_filter() {
    let mut system = System::new();
    let node0 = system
        .build_node()
        .config(NodeConfig {
            network_duplicate_filter_cutoff: 3, // Expire after 3 seconds
            ..System::default_config()
        })
        .finish();
    let node1 = system
        .build_node()
        .config(NodeConfig {
            network_duplicate_filter_cutoff: 3, // Expire after 3 seconds
            ..System::default_config()
        })
        .finish();

    let vote = Vote::new(&DEV_GENESIS_KEY, 0, 0, vec![*DEV_GENESIS_HASH]);
    let message = Message::ConfirmAck(ConfirmAck::new_with_own_vote(vote));

    // Publish duplicate detection through TCP
    let channel_id = node0
        .network_info
        .read()
        .unwrap()
        .find_node_id(&node1.node_id.public_key())
        .unwrap()
        .channel_id();

    // Send a vote
    node0.message_publisher.lock().unwrap().try_send(
        channel_id,
        &message,
        DropPolicy::ShouldNotDrop,
        TrafficType::Generic,
    );

    assert_always_eq(
        Duration::from_millis(100),
        || {
            node1.stats.count(
                StatType::Filter,
                DetailType::DuplicateConfirmAckMessage,
                Direction::In,
            )
        },
        0,
    );

    node0.message_publisher.lock().unwrap().try_send(
        channel_id,
        &message,
        DropPolicy::ShouldNotDrop,
        TrafficType::Generic,
    );

    assert_timely_eq(
        Duration::from_secs(2),
        || {
            node1.stats.count(
                StatType::Filter,
                DetailType::DuplicateConfirmAckMessage,
                Direction::In,
            )
        },
        1,
    );

    // The filter should expire the vote after some time
    let mut serializer =
        MessageSerializer::new(ProtocolInfo::default_for(Networks::NanoDevNetwork));
    let msg_bytes = serializer.serialize(&message);
    let payload_bytes = &msg_bytes[MessageHeader::SERIALIZED_SIZE..];
    assert!(node1.network_filter.check_message(&payload_bytes));
    assert_timely(Duration::from_secs(10), || {
        !node1.network_filter.check_message(&payload_bytes)
    });
}
